### DISCUSSION TITLE
Notification when multiple persons are working in 1 item

### DISCUSSION THREAD

**Discussion by DouweAVROTROS** [directus/directus#20594] [Feature Requests]
### Summary

When you edit at the same time, the same item. You don't know that someone else also is editing.

The risk is that person 1 something saves and person 2 overrides a few moments later (because he has an old version of the data).

There are 2 solutions where I'm thinking of:
1. Give a notification that someone else also is editing this item
2. Lock the item, so nobody can open it (not preferred, but some other CMS does this)

### Motivation

Our editors are overriding now each others work. They didn't knew that someone else also worked in the same item.

### Detailed Design

First, this is not waterproof. With a direct hit on the API, this won't work.

I also like that is configurable (like the preview url) per collection. To turn it on and of.

But to make it easier for a larger team to know what item is being edit right now, this can help.

I'm thinking about a notification on top of the item like:
> ❗ This item is being edit by: {{name}}

The second solution will be something like a icon of a lock in the list view 🔒 

### Must Have:
- Something to make it clear, that somebody is also editing right now
- Something to remove the notification when the user leaves

### Should Have:
- Status being cleared after a while

### Could Have:
- Realtime updates when somebody leaves the item

### Drawbacks

- It is not necessary for everybody
- It gives extra "bells and whistles" on Directus, what maybe not will be in line with the vision of Directus
- It will (probably) not work with the API
- The editors can also talk with each others first (why must Directus fix this 😉)

### Alternatives

Not doing this, is an option. It works as it is now.

Solution 2, with total locking is also a option. But not user friendly at all.

### Adoption Strategy

There will be no breaking changes.
There will probably a new field needed for locking, so maybe a migration is needed, but not on the core, just on the custom collections self. So it is then to every admin of the Directus CMS.

**Comment by BB-Loft** [comment]
On this Thursday Aug 1st, we're inviting you to come and have a chat with us about this feature request at our Request Review in our Discord server. This will help us have some dedicated time to talk about what implementation would have the most sense. Please feel free to join us there at https://directus.chat/ (the event is already listed and you can set a reminder).

**Reply by DouweAVROTROS** [reply]
Why was it rescheduled? We try to be there on sept 12.

**Reply by hanneskuettner** [reply]
Keen eye! We had some unavailabilities in our team so it had to be rescheduled to sept 12

**Reply by rijkvanzanten** [reply]
It was me! I had a minor surgery planned that got moved up an hour last minute. Didn't want to reschedule that out another couple months, so had to drop this weeks recording of request review 👍🏻

**Reply by DouweAVROTROS** [reply]
@BB-Loft @hanneskuettner  is it possible to reschedule to a different date, for example september 19th?
We have a teammeeting that day and can't be there on the 12th.

**Reply by BB-Loft** [reply]
Hey @DouweAVROTROS - usually we wouldn't reschedule a review once the date is posted but since we kicked this one back already, if the 26th September works better for you all, we can alter the topic on the 12th to be something else and move it to the next one which is every two weeks?

**Reply by DouweAVROTROS** [reply]
Hello @BB-Loft 26th September works for us. Thank you for the reschedule, we appreciate it.

**Comment by w0kyj** [comment]
## Request Review Notes
Discussion, feedback, and possible solutions for this feature.

### Possible Solutions
- Adding a notification that someone else is editing -- seems a straight forward feature to implement
- websockets could help, but not required
- where/how is indicator shown?
    - field
    - header
    - footer
    - information panel
    - other
- what is considered someone else editing
    - user with edit permissions is viewing the item
    - user with edit permissions makes a change, then they are editing
    - relational items -- must be considered for all cases
    - lock record when user begins edit
        - how long do you allow the lock to remain?
        - way to force release a lock
        - admin reset/force release
        - relational records? what happens with this
    - allow users to break another users lock?
- Extensions approach example with current websockets feature:
    - "presentational interface" extension to notify others have the current record open
    - https://github.com/directus-labs/extension-live-collaboration-interface
        - no locking involved just notifying
- Extensions options
    - add to data model field or collection with fields that auto track users opening items
    - flows or hooks to populate based on user navigation
    - update the notifier field that indicates someone else is in the item
    - ensure relational data is handled
        - mount and unmount when drawer opened
    - 
- implementation options
    - vim style, click a button to enter edit mode
        - could be done via extension using a toggle/manual flow button/...
        - permissions set to only allow updates when the value is toggled
        - see lock and relational data concerns above
        - 
    - automatic ping style where as long as user actively navigating interacting with item then it is "being edited", if the ping is idle for some rest period then it is no longer being edited
- 

### Best Approach
- Custom Extensions appears to be the best short term option
    - directus-labs submission to experiment with custom extensions
    - use this to identify more edge cases
    - relational handling

**Reply by wagnerSquid** [reply]
https://github.com/directus-labs/extension-live-collaboration-interface is a dead link. Is it anywhere else?

**Reply by rijkvanzanten** [reply]
Nah it's private atm. It was a quick and dirty proof of concept of a team member, but definitely not ready for primetime yet 🙂

**Reply by Abdallah-Awwad** [reply]
@wagnerSquid that's the new version most likely https://github.com/directus-labs/extensions/tree/main/packages/collaborative-editing

**Comment by MentalGear** [comment]
This is an essential feature to have. Maybe br41nslug who has done the websocket PR already can forward a PR for this one as well?

**Reply by rijkvanzanten** [reply]
Our current websocket support is all around subscribing to changes in collections, but doesn't have an open ended message bus with custom channels (yet). This type of notification would require a ephemeral channel to be opened for the page scope itself, with those messages not being tied to any particular collection 🤔

**Comment by mihaon** [comment]
I think the best approach is to use a "last update datetime" comparison.

When a user opens an item for editing:
1. Save opened item "last update datetime" (or last revision datetime) to frontend variable.

When a user saves an item:
1. Compare the old/new item "last update datetime" (or last revision datetime).
2. If the actual item "last update datetime" is not equal to frontend variable value then show a warning (maybe with a new values under the changed fields).
